ffi/src/owned.rs

```diff
@@ -0,0 +1,206 @@
+//! Logic related to the conversion of structures with ownership. Ownership is never transferred
+//! across FFI. This means that contents of these structures are copied into provided containers
+
+use crate::{
+    slice::{OutBoxedSlice, SliceRef},
+    AsReprCRef, FfiResult, IntoFfi, Output, ReprC, TryFromReprC,
+};
+
+/// Wrapper around `T` that is local to the conversion site. This structure carries
+/// ownership and care must be taken not to let it transfer ownership into an FFI function
+// NOTE: It's not possible to mutably reference local context
+#[derive(Clone)]
+#[repr(transparent)]
+pub struct Local<T>(pub(crate) T);
+
+/// Wrapper around [`Option<Box<[T]>>`] that is local to the conversion site. This structure
+/// carries ownership and care must be taken not to let it transfer ownership into FFI function
+// NOTE: It's not possible to mutably reference local context
+#[derive(Debug)]
+#[repr(C)]
+pub struct LocalSlice<T>(*mut T, usize);
```

Comment:
that's true, I think I left it like this on purpose because I wasn't sure that it can be covariant. I figured this is more safe and that it can always be expanded to covariance if necessary. Do you think there might be a reason why it couldn't be covariant?

Reply:
> Do you think there might be a reason why it couldn't be covariant?

Uff, it's hard to tell really, i will try to do a bit more research about this.

Another concern that i have is drop checker, we don't signal that we own `T`.
[Nomicom](https://doc.rust-lang.org/nomicon/phantom-data.html) has relevant example.
And if we look at [`Unique`](https://docs.rs/rust-libcore/latest/src/core/ptr.rs.html#981-989) pointer which is used inside `Box`, we can see that it uses `PhantomData<T>` to make it clear that owns the data.
Is this relevant to our case? What do you think?

Reply:
the ownership of data is a concern because of the drop check, I should add `PhantomData<T>` to `LocalSlice`

Reply:
> Uff, it's hard to tell really, i will try to do a bit more research about this.

I've made it covariant